geopandas/geodataframe.py

```diff
@@ -460,40 +465,36 @@ def dissolve(self, by=None, aggfunc='first'):
         aggfunc : function or string, default "first"
             Aggregation function for manipulation of data associated
             with each group. Passed to pandas `groupby.agg` method.
+        as_index : boolean, default True
+            If true, groupby columns become index of result.
 
         Returns
         -------
         GeoDataFrame
         """
 
         # Process non-spatial component
-        data = self.drop(labels=self.geometry.name, axis=1).copy()
+        data = self.drop(labels=self.geometry.name, axis=1)
         aggregated_data = data.groupby(by=by).agg(aggfunc)
 
 
         # Process spatial component
-        groupby_plus_geometry_cols = [self.geometry.name]
-        groupby_plus_geometry_cols.append(by)
-        geometry = self[groupby_plus_geometry_cols].copy()
-
         def merge_geometries(block):
-
             merged_geom = block.unary_union
+            return merged_geom
 
-            new_index = block.drop(self.geometry.name, axis=1).iloc[0][by]
-            merged_w_index = GeoSeries(merged_geom, index=Index(Series(new_index),name=by),
-                                          name=self.geometry.name)
-            return merged_w_index
-
+        g = self.groupby(by=by, group_keys=False)[self.geometry.name].agg(merge_geometries)
 
-        g = geometry.groupby(by=by, group_keys=False).apply(merge_geometries)
-
-        aggregated_geometry = GeoDataFrame(g,
-                                           index=g.index,
-                                           geometry=self.geometry.name)
+        # Aggregate
+        aggregated_geometry = GeoDataFrame(g, geometry=self.geometry.name)
         # Recombine
         aggregated = aggregated_geometry.join(aggregated_data)
         aggregated = aggregated.set_geometry(self.geometry.name)
```

Comment:
Normally, you should now be able to remove this line (as I merged my PR to fix this)